lib/ansible/plugins/vars/host_group_vars.py

```diff
@@ -55,60 +55,92 @@
 
 import os
 from ansible.errors import AnsibleParserError
-from ansible.module_utils.common.text.converters import to_bytes, to_native, to_text
+from ansible.module_utils.common.text.converters import to_bytes, to_native
 from ansible.plugins.vars import BaseVarsPlugin
-from ansible.inventory.host import Host
-from ansible.inventory.group import Group
+from ansible.utils.path import basedir
 from ansible.utils.vars import combine_vars
 
+CANONICAL_PATHS = {}
 FOUND = {}  # type: dict[str, list[str]]
+NAK = set()
+HOSTS = set()
+GROUPS = set()
+
+
+def load_found_files(loader, data, found_files):
+    for found in found_files:
+        new_data = loader.load_from_file(found, cache=True, unsafe=True)
+        if new_data:  # ignore empty files
+            data = combine_vars(data, new_data)
+    return data
 
 
 class VarsModule(BaseVarsPlugin):
 
     REQUIRES_ENABLED = True
+    reuse_instance = True
 
     def get_vars(self, loader, path, entities, cache=True):
         ''' parses the inventory file '''
 
         if not isinstance(entities, list):
             entities = [entities]
 
-        super(VarsModule, self).get_vars(loader, path, entities)
+        # realpath is expensive
+        if path not in CANONICAL_PATHS:
+            CANONICAL_PATHS[path] = os.path.realpath(basedir(path))
+        realpath_basedir = CANONICAL_PATHS[path]
 
         data = {}
         for entity in entities:
-            if isinstance(entity, Host):
+            try:
+                if not entity.name or entity.base_type not in ('Host', 'Group'):
+                    raise AttributeError
+            except AttributeError:
+                raise AnsibleParserError("Supplied entity must be Host or Group, got %s instead" % (type(entity)))
+
+            entity_id = entity.base_type + entity.name
+            if entity_id in HOSTS or entity_id in GROUPS:
+                subdir = 'host_vars' if entity_id in HOSTS else 'group_vars'
+            elif entity.base_type == 'Host':
                 subdir = 'host_vars'
-            elif isinstance(entity, Group):
+                HOSTS.add(entity_id)
+            elif entity.base_type == 'Group':
                 subdir = 'group_vars'
-            else:
-                raise AnsibleParserError("Supplied entity must be Host or Group, got %s instead" % (type(entity)))
+                GROUPS.add(entity_id)
 
             # avoid 'chroot' type inventory hostnames /path/to/chroot
-            if not entity.name.startswith(os.path.sep):
+            if entity.name[0] != os.path.sep:
```

Comment:
Yikes, I'd never really considered why this conditional is here- if we actually allow "pathy" looking things as host/group names by the time we've reached this point,  I suspect we might have a whole bunch of other "interesting" problems/bugs/potential-security-issues to solve...